### ISSUE TITLE
HDMI output defaulting to stereo in exclusive mode

### ISSUE THREAD

**Issue by stevenh** [mpv-player/mpv#10668]
### Important Information

Provide following Information:
- mpv 0.34.0
- Windows 10
- Build into Plex HTPC
- Issue introduced by https://github.com/mpv-player/mpv/commit/b2e5eb13bc08a0286782fed29455a66a037b46e5
- Possible screenshot or video of visual glitches

### Reproduction steps

Playback a multi channel audio file e.g. flac (sample attached) with a compatible HDMI output device.

### Expected behavior
Playback layout should match the file layout if supported by the HDMI output device.

### Actual behavior
Playback defaults to stereo.

### Log file
[htpc.log](https://github.com/mpv-player/mpv/files/9617918/htpc.log)

### Sample files
[sample.zip](https://github.com/mpv-player/mpv/files/9617930/sample.zip)

It's clear that this was an [intentional change](https://github.com/mpv-player/mpv/commit/b2e5eb13bc08a0286782fed29455a66a037b46e5), but looking at the log it seems that it should be possible to select a compatible layout as the enumeration shows it does know what channel formats are supported.

As the [commit which disables multichannel when exclusive mode is enabled](https://github.com/mpv-player/mpv/commit/b2e5eb13bc08a0286782fed29455a66a037b46e5) was 6 years ago, I'm wondering if its now possible to do better?

**Comment by mitzsch** [comment]
> Playback layout should match the file layout if supported by the HDMI output device.
With mpv and Plex HTPC this works. 
For Plex HTPC you need to select exclusive mode, HDMI for audio output and 7.1 for channel count, then it should work. (Auto does not work - This was different in Plex Media Player as far as I can remember it - Plex HTPC uses` -audio-channels=auto-safe`, whereas PMP used `-audio-channels=auto`)

For `-audio-channels=auto-safe ` have a lookt at the manual => https://mpv.io/manual/master/#options-audio-channels

For mpv run `mpv.com 7.1-multichannelfile.mkv -audio-exclusive=yes -audio-channels=auto` and it shoudl work.

**Comment by stevenh** [comment]
Yes enabling exclusive in Plex HTPC settings and then overriding `audio-channels`  in `mpv.conf` works, but it's not obvious, resulting in a poor user experience. You can see that in the [thread I created on the Plex forums](https://forums.plex.tv/t/no-high-res-audio-playback-using-htpc/811294), to get to this point.

So with that in mind, I'm wondering if that could be improved? From what I saw in the log mpv does know which layouts are supported, given this could that knowledge be used so that `audio-channels=auto-safe` works out the box?

The goal is to improve the end user experience, avoiding them having to manually config it.

Thoughts?

**Comment by mitzsch** [comment]
> Yes enabling exclusive in Plex HTPC settings and then overriding audio-channels in mpv.conf works, but it's not obvious, resulting in a poor user experience. 

Well, you don´t need to add it to the mpv.conf, setting it up in the Plex HTPC settings should be (is) sufficient. But yes not very user-friendly, I also needed some time to find the settings needed to get it working. (without the need of an mpv.conf)

I will fire up PlexMediaPlayer/Plex HTPC tomorrow on my HTPC to back up my statement regarding the audio-channel line. *
For PMP it was managed here => 
https://github.com/plexinc/plex-media-player/blob/d9685b274934855a707a37613cae09108ee9db6c/src/player/PlayerComponent.cpp#L1103


> So with that in mind, I'm wondering if that could be improved? From what I saw in the log mpv does know which layouts are supported, given this could that knowledge be used so that audio-channels=auto-safe works out the box?

Well, to get it working the plex devs need to change the regarding line to `audio-channels=auto`, then it should work. (Same experience as in PMP) I was never able to get the correct layout with `audio-channels=auto-safe`.

**Comment by garoto** [comment]
> poor user experience
> end user experience
> [...] not very user-friendly

mpv has no concept of "user-friendliness", whatever that means. One is supposed to keep track of the git commit history to have a mental resemblance of a "changelog" and adapt your .conf accordingly.

**Comment by stevenh** [comment]
> Well, you don´t need to add it to the mpv.conf, setting it up in the Plex HTPC settings should be (is) sufficient. But yes not very user-friendly, I also needed some time to find the settings needed to get it working. (without the need of an mpv.conf)

As far as I can tell there is no option in the HTPC settings that controls the `audio-channels`, the only option was adding an `mpv.conf`. Plex Media Player doesn't have this problem as the version of mpv used predates the mentioned changed so there was no `audio-channels=auto-safe` (the new default).

> Well, to get it working the plex devs need to change the regarding line to `audio-channels=auto`, then it should work. (Same experience as in PMP) I was never able to get the correct layout with `audio-channels=auto-safe`.

That would work but still doesn't fix it for any other mpv users. If mpv knows the supported layouts of the device then by definition they are safe, unless I'm missing something?

In this case improving the smarts of selection process so that it chooses from only those supported layouts would be the best route as it fixes it for everyone.

**Comment by mitzsch** [comment]
I dont know if it belongs here or better in the plex forum, but well now we are here and I will post it here.

> As far as I can tell there is no option in the HTPC settings that controls the audio-channels, the only option was adding an mpv.conf.

There is, have look at the pictures:
- Plex HTPC

![Screenshot 2022-09-22 095910-phtpc-7 1-settingspng](https://user-images.githubusercontent.com/43502191/191692442-f0d5f939-13a2-42ee-ac73-0b1287b51c6e.png)


![Screenshot 2022-09-22 095910-phtpc-7 1-cropped](https://user-images.githubusercontent.com/43502191/191692392-698b7019-9eed-451c-a9be-c56eb7a49470.png)

![Screenshot 2022-09-22 095910-phtpc-auto-settings](https://user-images.githubusercontent.com/43502191/191692527-5e7a834b-0de8-435b-853b-0ab29e066ad0.png)

![Screenshot 2022-09-22 095910-phtpc-auto-cropped](https://user-images.githubusercontent.com/43502191/191692808-ef8215f3-42a6-429e-8f4d-5dc7ae45a4b6.png)

----

- Plex Media Player

![Screenshot 2022-09-22 095407-pmp-7 1-settings](https://user-images.githubusercontent.com/43502191/191692965-aa143a9f-9ac9-4fee-bf53-950fe7a5f30e.png)

![Screenshot 2022-09-22 095407-pmp-7 1](https://user-images.githubusercontent.com/43502191/191693012-93bb7c73-427b-438d-b3f0-06786bcdf13b.png)

![Screenshot 2022-09-22 095407-pmp-auto-settings](https://user-images.githubusercontent.com/43502191/191693053-d42b9b17-1707-48b5-bee9-38e8e274b7e5.png)

![Screenshot 2022-09-22 095407-pmp-auto](https://user-images.githubusercontent.com/43502191/191693084-9016bbde-bdad-4634-85f4-bb8c69ff7685.png)



> That would work but still doesn't fix it for any other mpv users.

That's probably true, but mpv is not an out-of-the-box perfect media player it needs some tinkering by the user. There is a manual available, but yeah regarding this one it could have been written a bit clearer. So if you want to automatically set the output based on the file's audio layout you need to add `-audio-exclusive=yes -audio-channels=auto` to the command line or the mpv.conf.

**Comment by stevenh** [comment]
Thanks @mitzsch, you're correct there are workarounds however the goal of raising this issue is to improve mpv so any consumers of it will benefit.

I've checked the code our and will do some hacking and see what I can come up with.
